### ISSUE TITLE
nixos-rebuild test leaves around result link with a flake configuration

### ISSUE THREAD

**Issue by 2xsaiko** [NixOS/nixpkgs#170000]
### Describe the bug
Running `nixos-rebuild test` leaves a result symlink in the current directory. According to the man page, this is only supposed to happen with `nixos-rebuild build` and `build-vm`. However, it looks like for the flake builder, it only [doesn't place the result link into the current directory for nixos-rebuild switch and boot](https://github.com/NixOS/nixpkgs/blob/204366af6ddce48b873ad5b293fc61285e29a065/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L255).

### Steps To Reproduce
1. Have a flake system configuration
2. Run `nixos-rebuild test`
3. See result symlink in current directory

### Expected behavior
No result symlink is left around in the current directory.

### Notify maintainers

@thiagokokada 

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.111, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5.0pre20211206_d1aaa7e`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

**Comment by thiagokokada** [comment]
I have some commits in `nixos-rebuild` but I am no maintainer.

**Comment by lourkeur** [comment]
Looks like intended behavior based on the source code:

```
# Either upgrade the configuration in the system profile (for "switch"
# or "boot"), or just build it and create a symlink "result" in the
# current directory (for "build" and "test").
```

The reasoning seems to be that since `test` does not touch the profiles unlike the other commands, it must create its own garbage-collector root.

**Comment by 2xsaiko** [comment]
> I have some commits in `nixos-rebuild` but I am no maintainer.

Sorry! I looked at the 21.11 tree and there was no maintainer listed yet (should have looked in master) so I just picked the most recent committer.

> The reasoning seems to be that since `test` does not touch the profiles unlike the other commands, it must create its own garbage-collector root.

IMHO that is pretty annoying behavior, since I use `test` frequently to test small changes to my configuration, pretty often also from different directories, so I end up having result symlinks all over my configuration which I then have to manually remove.
Also, _does_ it have to create its own garbage-collector root? If you mean, to prevent accidentally collecting the currently running system, I assumed that that's what /nix/var/nix/gcroots/current-system is for, to prevent exactly that.

**Comment by lourkeur** [comment]
If the GC quicks in before `current-system` is set there is a race condition. I think `"${tmpDir}/result"` would be enough to prevent that though, but I haven't looked at it in-depth.

**Comment by 2xsaiko** [comment]
Yeah, that's what I'd think as well. I haven't looked too closely yet either, but I would assume that for switch/boot, the temp link stays around until all the operations are done, which means there's always at least one gc root to the new configuration.

**Comment by oati** [comment]
`nixos-rebuild dry-activate` also leaves behind a symlink

**Comment by LeviVanDerMaas** [comment]
Is this happening again for anyone else? This is happening again for me, at least as far back as nixpkgs unstable commit 30a61f056ac492e3b7cdcb69c1e6abdcf00e39cf

**Comment by thiagokokada** [comment]
> Is this happening again for anyone else? This is happening again for me, at least as far back as nixpkgs unstable commit [30a61f0](https://github.com/NixOS/nixpkgs/commit/30a61f056ac492e3b7cdcb69c1e6abdcf00e39cf)

Yes, it is a regression in `nixos-rebuild-ng`. Opened a PR to fix: https://github.com/NixOS/nixpkgs/pull/422490.
